src/backend/columnar/columnar_storage.c

```diff
@@ -0,0 +1,281 @@
+/*-------------------------------------------------------------------------
+ *
+ * columnar_storage.c
+ *
+ * This file contains function definitions for reading and writing to SMgr to
+ * logical offsets (ignoring page headers).
+ *
+ * Copyright (c) 2016, Citus Data, Inc.
+ *
+ * $Id$
+ *
+ *-------------------------------------------------------------------------
+ */
+
+
+#include "postgres.h"
+
+#include "safe_lib.h"
+
+#include "access/heapam.h"
+#include "access/nbtree.h"
+#include "catalog/pg_am.h"
+#include "miscadmin.h"
+#include "storage/fd.h"
+#include "storage/lmgr.h"
+#include "storage/smgr.h"
+#include "utils/guc.h"
+#include "utils/memutils.h"
+#include "utils/rel.h"
+#include "utils/relfilenodemap.h"
+
+#include "columnar/columnar.h"
+#include "columnar/columnar_storage.h"
+
+#define COLUMNAR_METAPAGE_BLOCKNO 0
+#define FIRST_LOGICAL_OFFSET COLUMNAR_BYTES_PER_PAGE
+#define LogicalOffsetIsValid(X) ((X) >= FIRST_LOGICAL_OFFSET)
+
+static void ReadFromBlock(Relation rel, BlockNumber blockno, uint32 offset,
+						  char *buf, uint32 len);
+static void WriteToBlock(Relation rel, BlockNumber blockno, uint32 offset,
+						 char *buf, uint32 len, bool clear);
+
+
+/*
+ * ColumnarMetapageInit - initialize a new metapage in an empty relation to
+ * the given contents.
+ */
+void
+ColumnarMetapageInit(Relation rel, ColumnarMetapage *metapage)
+{
+	LockRelationForExtension(rel, ExclusiveLock);
```

Comment:
I think it would be nice to have some simple isolation tests to show that concurrent calls made to `ColumnarMetapageInit` doesn't break anything.

It would be hard to explicitly show that one of the concurrent calls block on the other one but it would be nice if we can have a test udf to be used during those isolation tests to check if we have an healthy metapage in the end.

Reply:
concurrent calls made to `ReserveStripe` already block each other thanks to `LockForStripeReservation`, so it seems that the call made to `LockRelationForExtension` doesn't seem to be testable at the moment, so skip my comment.

Reply:
Implemented a udf to use in manual testing, sharing the commit if it makes sense to you to add some regression regression tests with it.

https://github.com/citusdata/citus/commit/db5600207ae42a1e8d9847d3598678649de4ade8

1) For example, we can have some simple tests like:

```sql
create table columnar_table(a int, b int) using columnar;
-- should return no rows
select * from print_columnar_metapage('columnar_table');

insert into columnar_table values (1);
-- see if metapage is as expected
select * from print_columnar_metapage('columnar_table');
....
```

2) If we decide to do those reservations in `ColumnarWriteRow`, than we should also have a few tests with roll-back'ed transactions.

3) After we decide how to handle columnar tables with old-style metapage/metadata, I think we should have some tests that call `print_columnar_metapage` before and after upgrade.

Reply:
resolving since https://github.com/citusdata/citus/pull/4907#pullrequestreview-638652251 is about the same thing